docs/2-Interviews/answering_technical_questions.md

```diff
@@ -1,39 +1,129 @@
+- Make sure you handle common edge cases, like an empty input! People always forget that
+
+A common question students ask is **"am I allowed to use library functions?"** (e.g. find element in array). The answer is **probably yes**. However:
+- If the library solves the question in one line, the interviewer probably wants you to implement it to see if you understand how it works. Feel free to mention that an implementation already exist, but just don't start complaining about how you wouldn't do it in real-life (it's not the point)
+- Do explain very clearly the input and outputs of the function you are using, and when possible how it's implemented or what kind of tricky behavior it might have
+- Just to make sure, it's worth asking permission, or at least make sure the interviewer knows what the function is
+- It's fine to make up a function if you can't remember what the existing one is called. Nobody cares if you use `File.open(...)` when it's supposed to be `Sys.open_file(...)`
+
+On that subject, a very useful bit of preparation that can only take an hour or two is to familiarize yourself with the libraries that come with your language. It's worth reading over the operations you can do on strings, arrays/lists and maps/dictionaries. Useful functions include `zip`, `split`, `substr`/`slice`, `join`, `find`, etc.
+
+### Algorithms: finishing the interview
+
+- Once you have an initial implementation mostly right, ask if any further optimizations are necessary
+  - Are you optimizing for time? Space? Which is more important?
+- Offer to make further improvements the code. This tends to impress interviewers
+  - How can the style of your code be improved?
+  - Is there any redudancy that could be eliminated?
+  - Could you refactor your solution to be more generalized? e.g. accept an arbitrary iterable data structure as an input when the question just asked you to handle arrays
+  - Can you write the solution more concisely? e.g. you used a bunch of for loops when you could have used map/filter/reduce
+- Write/talk through some test cases if you have time, to prove to yourself and the interviewer that your code works
+  - Volunteering to do this also impresses interviewers, because it shows that you take responsibility for making sure your code is correct
+
+### Algorithms: advanced tips
+
+- Implement functions in breadth-first order, not depth-first order
+  - For example, supposed you have your main function f1 that calls f2 and f3. Function f2 calls f4. An effective implementation order is f1, f2, f3, f4. Don't implement part of f1, jump to f2, then f3, and end up coming back to f1 later. You'll lose track of what you were doing. Instead, just call the functions, explaining to the interviewer what the function does and promise that you'll implement it after you're done with your current function
```

Comment:
This also optimizes the time you have. If you don't get to writing some of your helper functions that implement the details, the overall general solution will still be coded out in full